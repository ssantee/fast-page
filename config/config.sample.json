{
  "repositoryArn": "arn:aws:codecommit:us-west-1:123456789:my-repo",
  "configBucketName": "some-bucket-name",
  "tld": "example.com",
  "tldHzId": "Z02012345678",
  "environments": [
    {
      "name": "root",
      "domain": "fp.example.com",
      "adminDomain": "fpadmin.example.com",
      "apiDomain": "fpapi.example.com",
      "account": "123456789",
      "region": "us-west-1"
    },
    {
      "name": "dev",
      "domain": "dev.fp.example.com",
      "adminDomain": "dev.fpadmin.example.com",
      "apiDomain": "dev.fpapi.example.com",
      "account": "123456789",
      "region": "us-west-1",
      "pipelineRequiresApproval": false
    },
    {
      "name": "prod",
      "domain": "prod.fp.example.com",
      "adminDomain": "prod.fpadmin.example.com",
      "apiDomain": "prod.fpapi.example.com",
      "account": "123456789",
      "region": "us-west-1",
      "pipelineRequiresApproval": true
    }
  ],
  "parameterNames": {
    "certificateArn": "/fp/CertificateArn",
    "certificateArnAdmin": "/fp/adminCertificateArn",
    "certificateArnApi": "/fp/CertificateArnApi",
    "subdomainHostedZoneId": "/fp/subdomainHostedZoneId",
    "adminSubdomainHostedZoneId": "/fp/adminSubdomainHostedZoneId",
    "apiDomainHostedZoneId": "/fp/apidomainHostedZoneId",
    "ddbTableName": "/fp/ddbTable",
    "ddbTableArn": "/fp/ddbTableArn",
    "userPoolArn": "/fp/userPoolArn"
  }
}
