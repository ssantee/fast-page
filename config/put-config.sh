#!/bin/bash

BUCKET_NAME=$1

if [ -z "$BUCKET_NAME" ]; then
  echo "Arg 1 must be the bucket name."
  exit 1
fi

aws s3 cp ./config.json s3://${BUCKET_NAME}/config.json --profile default
aws s3 cp ../src/ui/admin/src/amplifyconfiguration.json s3://${BUCKET_NAME}/amplifyconfiguration.json --profile default
