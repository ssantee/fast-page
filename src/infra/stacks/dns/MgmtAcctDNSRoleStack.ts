import {
  aws_route53 as route53,
  aws_iam as iam,
  CfnOutput,
  Stack,
  StackProps,
  Duration,
} from "aws-cdk-lib";
import { Construct } from "constructs";
import { DomainEnv } from "../../../shared/types";

interface MgmtAcctDNSProps extends StackProps {
  iamPrincipalAccountNo: string;
  mgmtEnv: DomainEnv;
  targetEnv: DomainEnv;
  apiDomain: string;
}

export class MgmtAcctDNSRoleStack extends Stack {
  constructor(scope: Construct, id: string, props: MgmtAcctDNSProps) {
    super(scope, id);

    const newZone = new route53.PublicHostedZone(this, "HostedZone", {
      zoneName: props.mgmtEnv.domain,
    });

    const newAdminZone = new route53.PublicHostedZone(this, "HostedZoneAdmin", {
      zoneName: props.mgmtEnv.adminDomain,
    });

    const newApiZone = new route53.PublicHostedZone(this, "HostedZoneApi", {
      zoneName: props.apiDomain,
    });

    const crossAccountRole = new iam.Role(this, "CrossAccountRole", {
      // The role name must be predictable
      roleName: "MyXAcctDelegationRole",
      // The other account
      assumedBy: new iam.AccountPrincipal(props.targetEnv.account),
      // You can scope down this role policy to be least privileged.
      // If you want the other account to be able to manage specific records,
      // you can scope down by resource and/or normalized record names
      inlinePolicies: {
        crossAccountPolicy: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              sid: "ListHostedZonesByName",
              effect: iam.Effect.ALLOW,
              actions: ["route53:ListHostedZonesByName"],
              resources: ["*"],
            }),
            new iam.PolicyStatement({
              sid: "GetHostedZoneAndChangeResourceRecordSets",
              effect: iam.Effect.ALLOW,
              actions: [
                "route53:GetHostedZone",
                "route53:ChangeResourceRecordSets",
              ],
              // This example assumes the RecordSet subdomain.somexample.com
              // is contained in the HostedZone
              resources: [
                `arn:aws:route53:::hostedzone/${newZone.hostedZoneId}`,
              ],
              conditions: {
                "ForAllValues:StringLike": {
                  "route53:ChangeResourceRecordSetsNormalizedRecordNames": [
                    "dev.fp.santee.cloud",
                    "prod.fp.santee.cloud",
                    "fp.santee.cloud",
                    "fpadmin.santee.cloud",
                    "dev.fpadmin.santee.cloud",
                    "prod.fpadmin.santee.cloud",
                    "fpapi.santee.cloud",
                    "prod.fpapi.santee.cloud",
                    "dev.fpapi.santee.cloud",
                  ],
                },
              },
            }),
          ],
        }),
      },
    });
    newZone.grantDelegation(crossAccountRole);
    newAdminZone.grantDelegation(crossAccountRole);
    newApiZone.grantDelegation(crossAccountRole);

    const tldZone = route53.PublicHostedZone.fromPublicHostedZoneAttributes(
      this,
      "TLDZone",
      {
        hostedZoneId: "Z02022183L2XHLGIP6A83",
        zoneName: "santee.cloud",
      },
    );

    new route53.NsRecord(this, "SubDomainNSRecord", {
      zone: tldZone,
      recordName: props.mgmtEnv.domain,
      values: newZone.hostedZoneNameServers || [],
      ttl: Duration.seconds(172800),
    });

    new route53.NsRecord(this, "DomainNSRecordAdmin", {
      zone: tldZone,
      recordName: props.mgmtEnv.adminDomain,
      values: newAdminZone.hostedZoneNameServers || [],
      ttl: Duration.seconds(172800),
    });

    new route53.NsRecord(this, "ApiDomainNSRecord", {
      zone: tldZone,
      recordName: props.apiDomain,
      values: newApiZone.hostedZoneNameServers || [],
      ttl: Duration.seconds(172800),
    });

    new CfnOutput(this, "CrossAccountDelegationRoleArn", {
      description: "CrossAccountDelegationRoleArn",
      value: crossAccountRole.roleArn,
    });
  }
}
