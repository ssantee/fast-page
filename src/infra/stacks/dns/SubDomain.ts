import {
  aws_route53 as route53,
  aws_iam as iam,
  aws_ssm as ssm,
  Stack,
  StackProps,
  SecretValue,
  RemovalPolicy,
} from "aws-cdk-lib";
import { Construct } from "constructs";
import { DomainEnv } from "../../../shared/types";
import {
  Certificate,
  CertificateValidation,
} from "aws-cdk-lib/aws-certificatemanager";
import { Secret } from "aws-cdk-lib/aws-secretsmanager";

interface SubDomainProps extends StackProps {
  mgmtEnvAcctNo: string;
  certificateArnParamName: string;
  subdomainHostedZoneIdParamName: string;
  deployEnvDomain: string;
  mgmtEnvDomain: string;
}

export class SubDomain extends Stack {
  constructor(scope: Construct, id: string, props: SubDomainProps) {
    super(scope, id);

    const subZone = new route53.PublicHostedZone(this, "Route53FPSubZone", {
      zoneName: props.deployEnvDomain,
    });

    // import the delegation role by constructing the roleArn
    const delegationRoleArn = Stack.of(this).formatArn({
      region: "", // IAM is global in each partition
      service: "iam",
      account: props.mgmtEnvAcctNo,
      resource: "role",
      resourceName: "MyXAcctDelegationRole", // Do not change
    });

    const delegationRole = iam.Role.fromRoleArn(
      this,
      "DelegationRole",
      delegationRoleArn,
    );

    // create the record
    new route53.CrossAccountZoneDelegationRecord(
      this,
      "CrossAccountZoneDelegationRecord",
      {
        delegatedZone: subZone,
        parentHostedZoneName: props.mgmtEnvDomain,
        delegationRole,
        assumeRoleRegion: "us-east-1",
      },
    );

    const certificate = new Certificate(this, "DelegatedZoneCert", {
      domainName: props.deployEnvDomain,
      validation: CertificateValidation.fromDns(subZone),
    });

    new ssm.StringParameter(this, "mySsmParameter", {
      parameterName: props.certificateArnParamName,
      stringValue: certificate.certificateArn,
    });

    new ssm.StringParameter(this, "mySsmParameter1", {
      parameterName: props.subdomainHostedZoneIdParamName,
      stringValue: subZone.hostedZoneId,
    });
  }
}
