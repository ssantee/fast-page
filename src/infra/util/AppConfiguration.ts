export type EnvConfig = {
  name: string;
  domain: string;
  account: string;
  region: string;
  adminDomain: string;
  apiDomain: string;
};

type ParamConfig = {
  certificateArn: string;
  certificateArnAdmin: string;
  certificateArnApi: string;
  subdomainHostedZoneId: string;
  adminSubdomainHostedZoneId: string;
  apiDomainHostedZoneId: string;
  ddbTableName: string;
  ddbTableArn: string;
};

export type AppConfig = {
  environments: EnvConfig[];
  parameterNames: ParamConfig;
};

export class AppConfiguration {
  public deployEnv: string;
  public targetEnv: EnvConfig;
  public mgmtEnv: EnvConfig;
  public paramNames: ParamConfig;

  constructor(configData: AppConfig, deployEnv: string) {
    this.deployEnv = deployEnv;
    this.targetEnv = this.getTargetEnv(configData.environments);
    this.mgmtEnv = this.getMgmtEnv(configData.environments, "root");
    this.paramNames = configData.parameterNames;
  }

  private getTargetEnv(envs: EnvConfig[]): EnvConfig {
    return envs.find((e) => e.name === this.deployEnv)!;
  }
  private getMgmtEnv(envs: EnvConfig[], mgmtAcctName: string): EnvConfig {
    return envs.find((e) => e.name === mgmtAcctName)!;
  }
}
