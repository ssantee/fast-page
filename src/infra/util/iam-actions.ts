/*
 * Store the various IAM action strings used in the project.
 * */
enum ddbPermissions {
  PutItem = "dynamodb:PutItem",
  Scan = "dynamodb:Scan",
  GetItem = "dynamodb:GetItem",
  UpdateItem = "dynamodb:UpdateItem",
  DeleteItem = "dynamodb:DeleteItem",
}

enum s3Permissions {
  PutObject = "s3:PutObject",
  GetObject = "s3:GetObject",
  DeleteObject = "s3:DeleteObject",
  ListObjects = "s3:ListObjects",
}

export { ddbPermissions, s3Permissions };
