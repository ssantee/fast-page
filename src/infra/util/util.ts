import { Fn, Stack } from "aws-cdk-lib";

/*
 * getSuffixFromStack is used to get the suffix from the stackId.
 * Use this in naming resources like buckets and tables.
 * There are two benefits:
 * 1. It reduces the chances of name collisions.
 * 2. It gives an at-a-glance idea of which stack the resource belongs to.
 * */
export function getSuffixFromStack(stack: Stack) {
  const shortStackId = Fn.select(2, Fn.split("/", stack.stackId));
  return Fn.select(4, Fn.split("-", shortStackId));
}
