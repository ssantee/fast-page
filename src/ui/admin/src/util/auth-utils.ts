import { runWithAmplifyServerContext } from "@/util/server-utils";
import { AmplifyServer } from "aws-amplify/adapter-core";

// The fetchAuthSession is pulled as the server version from aws-amplify/auth/server
import { fetchAuthSession } from "aws-amplify/auth/server";
import { NextRequest, NextResponse } from "next/server";

export default async function checkAuth(
  request: NextRequest,
  response: NextResponse,
) {
  return await runWithAmplifyServerContext({
    nextServerContext: { request, response },
    operation: async (contextSpec: AmplifyServer.ContextSpec) => {
      try {
        // The fetch will grab the session cookies
        const session = await fetchAuthSession(contextSpec, {});
        console.log("session");
        console.log(session);
        return session.tokens !== undefined;
      } catch (error) {
        console.log(error);
        return false;
      }
    },
  });
}
