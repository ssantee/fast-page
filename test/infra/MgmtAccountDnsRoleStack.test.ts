import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { MgmtAcctDNSRoleStack } from "../../src/infra/stacks/dns/MgmtAcctDNSRoleStack";
import { describe } from "node:test";
import { AppConfiguration } from "../../src/infra/util/AppConfiguration";
import * as appConfig from "../../config/config.json";

describe("MgmtAcctDNSRoleStack test suite", () => {
  let MgmtAcctDNSRoleStackTemplate: Template;
  const appCfg = new AppConfiguration(appConfig, "prod");
  const targetEnv = appCfg.targetEnv;
  const mgmtEnv = appCfg.mgmtEnv;

  beforeAll(() => {
    const app = new cdk.App();
    const mgmtAcctDNSRoleStack = new MgmtAcctDNSRoleStack(
      app,
      "WebPublicStack",
      {
        env: {
          account: mgmtEnv.account,
          region: mgmtEnv.region,
        },
        mgmtEnv: mgmtEnv,
        targetEnv: targetEnv,
        iamPrincipalAccountNo: mgmtEnv.account,
        apiDomain: mgmtEnv.apiDomain,
      },
    );
    MgmtAcctDNSRoleStackTemplate = Template.fromStack(mgmtAcctDNSRoleStack);
  });

  test("creates role with correct role props", () => {
    MgmtAcctDNSRoleStackTemplate.hasResourceProperties("AWS::IAM::Role", {
      RoleName: "MyXAcctDelegationRole",
      AssumeRolePolicyDocument: {
        Statement: [
          {
            Effect: "Allow",
            Principal: {
              AWS: {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      Ref: "AWS::Partition",
                    },
                    `:iam::${targetEnv.account}:root`,
                  ],
                ],
              },
            },
          },
        ],
      },
    });
  });

  test("creates ns record with props", () => {
    MgmtAcctDNSRoleStackTemplate.hasResourceProperties(
      "AWS::Route53::RecordSet",
      {
        Name: `${mgmtEnv.domain}.`,
        Type: "NS",
      },
    );
  });

  test("Matches snapshot", () => {
    expect(MgmtAcctDNSRoleStackTemplate).toMatchSnapshot();
  });
});
