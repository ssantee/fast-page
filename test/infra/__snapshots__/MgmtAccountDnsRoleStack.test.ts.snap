// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
{
  "Outputs": {
    "CrossAccountDelegationRoleArn": {
      "Description": "CrossAccountDelegationRoleArn",
      "Value": {
        "Fn::GetAtt": [
          "CrossAccountRoleFACE29D1",
          "Arn",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ApiDomainNSRecord5AE8A5F1": {
      "Properties": {
        "HostedZoneId": "Z02022183L2XHLGIP6A83",
        "Name": "fpapi.santee.cloud.",
        "ResourceRecords": {
          "Fn::GetAtt": [
            "HostedZoneApiE1F610D8",
            "NameServers",
          ],
        },
        "TTL": "172800",
        "Type": "NS",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "CrossAccountRoleDefaultPolicy212A317F": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "route53:ChangeResourceRecordSets",
              "Condition": {
                "ForAllValues:StringEquals": {
                  "route53:ChangeResourceRecordSetsActions": [
                    "UPSERT",
                    "DELETE",
                  ],
                  "route53:ChangeResourceRecordSetsRecordTypes": [
                    "NS",
                  ],
                },
              },
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":route53:::hostedzone/",
                    {
                      "Ref": "HostedZoneDB99F866",
                    },
                  ],
                ],
              },
            },
            {
              "Action": "route53:ListHostedZonesByName",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "route53:ChangeResourceRecordSets",
              "Condition": {
                "ForAllValues:StringEquals": {
                  "route53:ChangeResourceRecordSetsActions": [
                    "UPSERT",
                    "DELETE",
                  ],
                  "route53:ChangeResourceRecordSetsRecordTypes": [
                    "NS",
                  ],
                },
              },
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":route53:::hostedzone/",
                    {
                      "Ref": "HostedZoneAdminA7C7F609",
                    },
                  ],
                ],
              },
            },
            {
              "Action": "route53:ChangeResourceRecordSets",
              "Condition": {
                "ForAllValues:StringEquals": {
                  "route53:ChangeResourceRecordSetsActions": [
                    "UPSERT",
                    "DELETE",
                  ],
                  "route53:ChangeResourceRecordSetsRecordTypes": [
                    "NS",
                  ],
                },
              },
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":route53:::hostedzone/",
                    {
                      "Ref": "HostedZoneApiE1F610D8",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CrossAccountRoleDefaultPolicy212A317F",
        "Roles": [
          {
            "Ref": "CrossAccountRoleFACE29D1",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CrossAccountRoleFACE29D1": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::590184120807:root",
                    ],
                  ],
                },
              },
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::339712782235:root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "route53:ListHostedZonesByName",
                  "Effect": "Allow",
                  "Resource": "*",
                  "Sid": "ListHostedZonesByName",
                },
                {
                  "Action": [
                    "route53:GetHostedZone",
                    "route53:ChangeResourceRecordSets",
                  ],
                  "Condition": {
                    "ForAllValues:StringLike": {
                      "route53:ChangeResourceRecordSetsNormalizedRecordNames": [
                        "fp.santee.cloud",
                        "fpadmin.santee.cloud",
                        "fpapi.santee.cloud",
                        "dev.fp.santee.cloud",
                        "dev.fpadmin.santee.cloud",
                        "dev.fpapi.santee.cloud",
                        "prod.fp.santee.cloud",
                        "prod.fpadmin.santee.cloud",
                        "prod.fpapi.santee.cloud",
                      ],
                    },
                  },
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:route53:::hostedzone/",
                        {
                          "Ref": "HostedZoneDB99F866",
                        },
                      ],
                    ],
                  },
                  "Sid": "GetHostedZoneAndChangeResourceRecordSets",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "crossAccountPolicy",
          },
        ],
        "RoleName": "MyXAcctDelegationRole",
      },
      "Type": "AWS::IAM::Role",
    },
    "DomainNSRecordAdmin9B44FC4B": {
      "Properties": {
        "HostedZoneId": "Z02022183L2XHLGIP6A83",
        "Name": "fpadmin.santee.cloud.",
        "ResourceRecords": {
          "Fn::GetAtt": [
            "HostedZoneAdminA7C7F609",
            "NameServers",
          ],
        },
        "TTL": "172800",
        "Type": "NS",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "HostedZoneAdminA7C7F609": {
      "Properties": {
        "Name": "fpadmin.santee.cloud.",
      },
      "Type": "AWS::Route53::HostedZone",
    },
    "HostedZoneApiE1F610D8": {
      "Properties": {
        "Name": "fpapi.santee.cloud.",
      },
      "Type": "AWS::Route53::HostedZone",
    },
    "HostedZoneDB99F866": {
      "Properties": {
        "Name": "fp.santee.cloud.",
      },
      "Type": "AWS::Route53::HostedZone",
    },
    "SubDomainNSRecordBDFBB65E": {
      "Properties": {
        "HostedZoneId": "Z02022183L2XHLGIP6A83",
        "Name": "fp.santee.cloud.",
        "ResourceRecords": {
          "Fn::GetAtt": [
            "HostedZoneDB99F866",
            "NameServers",
          ],
        },
        "TTL": "172800",
        "Type": "NS",
      },
      "Type": "AWS::Route53::RecordSet",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
